module2/lessons/serializers.md

```diff
@@ -0,0 +1,218 @@
+
+You can view the docs on the jsonapi-serializer gem [here](https://github.com/jsonapi-serializer/jsonapi-serializer#installation).
+
+Add it to your Gemfile
+
+```ruby
+gem "jsonapi-serializer"
+```
+
+And go ahead and
+
+```bash
+$ bundle install
+```
+
+This gem gives us a built in generator to make ourselves all of the serializers we could possibly want.
+
+```bash
+$ rails g serializer Song title
+```
+
+We can take a look and see what’s inside it.
+
+*app/serializers/store_serializer.rb*
+
+```ruby
+class SongSerializer
+  include JSONAPI::Serializer
+  attributes :title
+end
+```
+
+Now that we have our serializer, we need to edit our controller to use said serializer.
+
+*app/controllers/api/v1/songs_controller.rb*
+
+```ruby
+class Api::V1::SongsController < ApplicationController
+  def index
+    render json: SongSerializer.new(Song.all)
+  end
+
+  def show
+    render json: SongSerializer.new(Song.find(params[:id]))
+  end
+end
+```
+
+We can see that instead of just letting our ActiveRecord be rendered in JSON, we are going to take our collections, and send them to the serializer, and have the result of THAT be converted to JSON.
+
+This Gem can also be used to easily set up relationships between entities. You will use this functionality later in the mod!
+
+<!-- 
+This syntax is a bit different from what we are used to. We use `attribute` singular, and then as a symbol we pick the name of what we want our attribute to be. We use a do end block similar to an enumerable with a block parameter. Now the block parameter, `object` is a lot like self. We get to use it for each single thing of a collection we pass to the serializer. We are essentially saying for each thing you serialize, grab the books and count them too. In this manner we can add a custom generated value for each book.
+
+We can also have a custom static attribute like so:
+
+*app/serializers/store_serializer.rb*
+
+```ruby
+class StoreSerializer
+  include JSONAPI::Serializer
+  attributes :name
+
+  has_many :books
+
+  attribute :num_books do |object|
+    object.books.count
+  end
+
+  attribute :active do
+    true
+  end
+end
+```
+
+Alternatively we could create a num_books method in our `Store` model and then set it as an attribute in our serializer:
+
+*app/models/store.rb*
+
+```ruby
+class Store < ApplicationRecord
+  has_many :store_books 
+  has_many :books, through: :store_books
+  
+  def num_books
+    self.books.count
+  end 
+end
+```
+
+*app/serializers/store_serializer.rb*
+
+```ruby
+class StoreSerializer
+  include JSONAPI::Serializer
+  attributes :name, :num_books
+
+  has_many :books
+  
+  attribute :active do
+    true
+  end
+end
+```
+ -->
+
+<!-- Completed version of this lesson to this point is available [here](https://github.com/turingschool-examples/building_internal_apis_7/tree/customizing-json). -->
+
+## Extra Practice
+
+<!-- TODO, not sure what extra practice we want because we just have one model here... -->
```

Comment:
I would like some additional practice here, but with only one model I'm unsure what to do for an extension. Any ideas?